pkg/metallb/frr.go

```diff
@@ -0,0 +1,548 @@
+	return schema.GroupVersionResource{
+		Group: FRRAPIGroup, Version: APIVersion, Resource: "frrconfigurations",
+	}
+}
+
+// Exists checks whether the given FrrConfiguration exists.
+func (builder *FrrConfigurationBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if FrrConfiguration %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// WithBGPRouter defines a BGP router as localASN.
+func (builder *FrrConfigurationBuilder) WithBGPRouter(localASN uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP router with an ASN: %v", localASN)
+
+	if strconv.Itoa(int(localASN)) == "" {
+		glog.V(100).Infof("The local ASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'local AS' cannot be empty"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Iterate over the existing routers to find a match
+	for i := range builder.Definition.Spec.BGP.Routers {
+		if builder.Definition.Spec.BGP.Routers[i].ASN == localASN {
+			// Return the existing router
+			return builder
+		}
+	}
+
+	// If no router with the given ASN exists, create a new one
+	builder.Definition.Spec.BGP.Routers = append(builder.Definition.Spec.BGP.Routers, frrtypes.Router{
+		ASN: localASN,
+	})
+
+	return builder
+}
+
+// WithToReceiveModeAll allows all incoming routes.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeAll() *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive all routes in namespace %s with mode type: all",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Check if the BGP routers exist
+	if len(builder.Definition.Spec.BGP.Routers) == 0 {
+		builder.errorMsg = "No BGP routers defined."
+
+		return builder
+	}
+
+	// Iterate through routers to find the one with the desired ASN (if applicable) and update its neighbors
+	for i, router := range builder.Definition.Spec.BGP.Routers {
+		for j := range router.Neighbors {
+			// Update the ToReceive mode to AllowAll for each neighbor
+			builder.Definition.Spec.BGP.Routers[i].Neighbors[j].ToReceive.Allowed.Mode = frrtypes.AllowAll
+		}
+	}
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(prefixes []string,
+	neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive specific routes in namespace %s in mode type: filtered",
+		builder.Definition.Namespace)
+
+	if builder.errorMsg != "" {
```

Comment:
this validation should be after all validations and remove "return builder" in all vaildation then

Reply:
updated